### ISSUE TITLE
Would it be possible to create some sort of simple API or interface so anyone can interact from outside with DosBox?

### ISSUE THREAD

**Issue by HeuristicPerson** [dosbox-staging/dosbox-staging#904]
Would it be possible to create some sort of simple API or interface so anyone can interact from outside with DosBox?

That way we can not only change the configuration but also send commands from external programs to launch programs inside DosBox, set environment variables and so on...

_Originally posted by @HeuristicPerson in https://github.com/dosbox-staging/dosbox-staging/issues/315#issuecomment-782712236_

**Comment by dreamer** [comment]
Good idea - @HeuristicPerson do you have some specific API in mind? Right our API "surface" is sum of all available startup switches, .conf file parameters and accepted environment variables… but those can be set either via .conf file or on startup.

I believe you rather want something to interact with emulator while it's running, even if a DOS program is already started?

**Comment by kcgen** [comment]
An easy surface would be string-entry access to the command line itself - because it's already a user interface, behind which configuration options are parsed, commands are parsed, and so on.
It would give the API ongoing an identical access as one has on the keyboard - without having to maintain a giant suite of API calls.

**Comment by dreamer** [comment]
The idea I was toying in my head recently was: 

- we already have an ability to place fake programs on Z:\> drive and create "fake" DOS commands
- and we know that our DOS emulation capabilities are good enough support even quite complicated TRS programs…
- so maybe we can create a "virtual" builtin TRS program (virtual with the purpose of not occupying emulator's memory), and trigger it via ordinary keyboard shortcut (e.g. Ctrl+F1 or Ctrl+F2); the effect would be: currently running game disappears and our virtual TRS text UI is shown instead. When user closes TUI interface, TRS switches places with emulated games and user can continue as normal.

Architecturally, this would be very different than current Ctrl+f1 keymapper interface (which is a separate window with a separate event loop).

*I have no idea if this would work in practice or even if it's a good path forward.*

**Comment by HeuristicPerson** [comment]
> An easy surface would be string-entry access to the command line itself - because it's already a user interface, behind which configuration options are parsed, commands are parsed, and so on.

...plus the ability to send keystrokes because for something like increasing cycles, taking screenshots and so on are not command driven. Something like that, and printing the output to stdout should cover any case I can think of.  The reason I came to this feature request is because I was thinking about building a physical interactive panel based on arduino to control DosBox.

I think .conf files are not enough from the usability point of view since they force you to restart DosBox to change some settings or to directly type the commands in Dos (e.g. changing environment variables to activate/deactivate Soundblaster, GUS...). Another example is the automation of mounting floppy images. There is no easy way to create a DosBox batch file to automate the task because for loops are not implemented (yet?). An external api would allow you to have external scripts written in any modern language to read all the images in an specific folder and build the proper DosBox command line to mount all of them. I think the posibilities are endless and you wouldn't need to modify DosBox "core" code to add new extra features.

Regards

**Comment by kcgen** [comment]
@dreamer , the TSR approach sounds very powerful; it's API breadth could be as wide as the possibilities of what TSRs can do in DOS.

This would also acts as a simplified abstraction for non-DOS programmers. For example, take myself: I have no idea how to write a TSR cheat tool in DOS, but I do understand the concept of scanning memory values for changes. So if an API like that were provided, then it would open doors to those who aren't niche experts.

I also like that this would be active in parallel to a program or game. 

The command line API would be the most simple form of access, but still be quite powerful in terms of mechazing "configure and control" before and after a DOS program runs.  So I see unique value in both approaches.

**Comment by bleuge** [comment]
Maybe someone find this interesting: expose via Api the debugger, or part of it, showing cpu state, memory, ports, ability to step, and thats all. 
In this way, anyone could create a gui/tui interface for debugging, scriptable, etc. Via an external program.

**Comment by dreamer** [comment]
@bleuge unfortunately, it's not as easy as it sounds (debugger is extremely invasive and slows down emulation, which is unacceptable for normal use-cases - that's why debugger can be turned on only during compilation time). The issue for discussing it is #402 - if there's anyone willing to work on it, then feel free to contribute :)

This issue is about creating an interface to interact with DOS shell from outside of the emulator - which is a somewhat different usecase than hooking into the built-in debugger.

**Comment by kcgen** [comment]
Although very interesting, this is quite outside the scope of Staging.  If someone can write a PR and it's elegant and does not harm performance or maintainability - then great.  Closing this for now, as it's not in scope.

**Comment by QmwJlHuSg9pa** [comment]
Hi, can anyone suggest a solution for emulating keystrokes from within DOS/DOSBox? I want to send single key inputs to a DOS setup program (in order to automate video/audio selection, etc.), and explicitly set pauses of arbitrary duration between each key input.

**Comment by johnnovak** [comment]
> Hi, can anyone suggest a solution for emulating keystrokes from within DOS/DOSBox? I want to send single key inputs to a DOS setup program (in order to automate video/audio selection, etc.), and explicitly set pauses of arbitrary duration between each key input.

Not possible, and out of scope.

**Comment by kcgen** [comment]
We don't have a bulk input method, however for just a handful of keystrokes `autotype` does the job. Read about it here w/ examples: https://github.com/dosbox-staging/dosbox-staging/wiki/AUTOTYPE-Candidates

**Comment by QmwJlHuSg9pa** [comment]
> We don't have a bulk input method, however for just a handful of keystrokes `autotype` does the job. Read about it here w/ examples: https://github.com/dosbox-staging/dosbox-staging/wiki/AUTOTYPE-Candidates

This is exactly what I was looking for, thank you.
